mdx/guides/develop-a-margin-trading-smart-contract-with-0x-api.mdx

```diff
@@ -0,0 +1,573 @@
+
+Once verified, let’s perform the swap:
+
+```solidity
+// 3. execute swap
+(bool success, bytes memory data) = address(EXCHANGE).call.value(quote.protocolFee)(quote.calldataHex);
+require(success, "Swap not filled.");
+// 4. decode results
+LibFillResults.FillResults memory fillResults = abi.decode(data, (LibFillResults.FillResults));
+```
+
+At this point, the contract should have a balance of `daiBorrowBalance` DAI held by the contract.
+
+#### Repay borrowed DAI and withdraw ETH
+
+With the DAI bought with 0x API, let’s repay back our `borrowBalance` on Compound finance:
+
+```solidity
+...
+// 5. return back DAI
+_approve(address(DAI), address(CDAI));
+require(CDAI.repayBorrow(fillResults.makerAssetFilledAmount) == 0);
+...
+```
+
+To return DAI back, an approval for the `CDAI` contract is needed. With `fillResults.makerAssetFilledAmount` we can repay the exact amount of DAI acquired with 0x. Once, `borrowBalance` is repaid, let’s withdraw the ETH deposited in `CETH`.
+
+```solidity
+...
+require(CETH.redeem(CETH.balanceOfUnderlying(address(this))) == 0);
+...
+```
+
+At this point, the contract holds ETH and WETH (whatever WETH remains not swapped for DAI). For the sake of UX, we will be unwrapping WETH to ETH and transferring all the ETH back to the owner. Then let’s reset `positionBalance` and allow another position to be opened.
+
+```solidity
+...
+    // 7. withdraw all WETH Balance;
+    WETH.withdraw(wethBalance);
+    // 8. transfer all ETH back to owner;
+    ethBalance = address(this).balance;
+    OWNER.transfer(address(this).balance);
+    // 9. reset balance
+    positionBalance = 0;
+}
+...
+```
+
+At this point `close` function should look something like this:
+
+```solidity
+function close(
+    ZeroExQuote memory quote
+)
+    public
+    onlyOwner
+    onlyWhenOpen
+    returns (uint ethBalance)
+{
+    // approve for swap
+    _approve(address(WETH), _getZeroExApprovalAddress());
+    // verify swap
+    uint256 wethBalance = WETH.balanceOf(address(this));
+    uint256 daiBorrowBalance = CDAI.borrowBalanceCurrent(address(this)); // TODO
+
+    require(wethBalance > quote.sellAmount, "not enough to swap");
+    require(quote.buyToken == address(DAI), "not buying DAI");
+    require(daiBorrowBalance < quote.buyAmount, "not enough DAI to repay");
+    // execute swap
+    (bool success, bytes memory data) = address(EXCHANGE).call.value(quote.protocolFee)(quote.calldataHex);
+    require(success, "Swap not filled.");
+    // decode results
+    LibFillResults.FillResults memory fillResults = abi.decode(data, (LibFillResults.FillResults));
+    // return back DAI
+    _approve(address(DAI), address(CDAI));
+    // approve Compound Finance CDAI
+    assert(CDAI.repayBorrow(fillResults.makerAssetFilledAmount) == 0);
```

Comment:
As stated above, let's use `require` (unless I'm missing context), and add an error message.